### PR TITLE
fix mqa preshuffle layout issue for deepseek v4

### PR DESCRIPTION
**Pull request by 13524182838**
## Motivation

AITER's preshuffle path is a paired writer/reader contract:

Writer: indexer_k_quant_and_cache_kernel(..., preshuffle=True) or an
equivalent writer must place FP8 K bytes in MFMA 16x16 preshuffled layout.
Reader: deepgemm_fp8_paged_mqa_logits(..., Preshuffle=True) or
cp_gather_indexer_k_quant_cache(..., preshuffle=True) reads that same layout.
Scale layout is unchanged: one FP32 scale per token lives after the
block_size * head_dim K-byte region for each block.
For a logical K element (token = block_offset, dim = head_dim_idx), preshuffle
stores it at:

token_tile_id = block_offset / 16
token_in_tile = block_offset % 16
col_tile_id = head_dim_idx / 16
col_in_tile = head_dim_idx % 16

offset =
  block_idx * cache_block_size * cache_stride
+ token_tile_id * (16 * head_dim)
+ col_tile_id * (16 * 16)
+ token_in_tile * 16
+ col_in_tile
Current SGLang Writer Layout
The current fused-store indexer writer appears row-major:

File: sglang/sglang/python/sglang/jit_kernel/triton_store_cache.py
Kernel: _triton_fused_store_indexer_kernel
Around lines 178-186
fp8_offset = page * BYTES_PER_PAGE + slot * HEAD_DIM + lane
The CUDA/C++ JIT version is also row-major:

File: sglang/sglang/python/sglang/jit_kernel/csrc/deepseek_v4/store.cuh
Kernel: fused_store_indexer_cache
Around lines 109-118
value_ptr = page_ptr + offset * 128
scale_ptr = page_ptr + 128 * page_size + offset * 4
The fallback writer also appears row-major:

File: sglang/sglang/python/sglang/srt/layers/attention/dsv4/index_buf_accessor.py
Kernel: _set_k_and_s_triton_kernel
Around lines 160-185
out_k_nope_offsets =
  loc_page_index * BUF_NUMEL_PER_PAGE
+ loc_token_offset_in_page * NUM_NOPE_ROPE_BYTES_PER_TOKEN
+ nope_range

## Modifications

Add related preshuffle layout for mqa in hip kernel(indexer).

## Accuracy Tests

accuracy 0.95 for GSM8K.

## Speed Tests and Profiling

It is just related to address of layout. No obvious performance loss.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30146451138](https://github.com/sgl-project/sglang/actions/runs/30146451138)
Latest PR Test (Extra): :x: [Run #30146451097](https://github.com/sgl-project/sglang/actions/runs/30146451097)

### CODE REVIEW

File: python/sglang/jit_kernel/csrc/deepseek_v4/fused_norm_rope_v2.cuh [hunk 1]
```diff
@@ -213,7 +213,19 @@ INDEXER_KERNEL void fused_norm_rope_indexer(const __grid_constant__ FusedNormRop
...
-    result.store(value_ptr, lane_id);
+    if constexpr (kPreshuffleSize != 0) {
+      constexpr int32_t kTile = kPreshuffleSize;
+      const int32_t dim_base = lane_id * kVecSize;
+      const int32_t token_tile_id = offset / kTile;
+      const int32_t token_in_tile = offset % kTile;
+      const int32_t col_tile_id = dim_base / kTile;
+      const int32_t col_in_tile = dim_base % kTile;
+      const int32_t value_offset =
+          token_tile_id * (kTile * kHeadDim) + col_tile_id * (kTile * kTile) + token_in_tile * kTile + col_in_tile;
```
[the remark this answers is not in the document]
Comment (13524182838) [reply python/sglang/jit_kernel/csrc/deepseek_v4/fused_norm_rope_v2.cuh]: I resolved it in the latest version.

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 2]
```diff
@@ -153,7 +154,7 @@ def _aiter_fp8_paged_mqa_logits(
         page_table.to(torch.int32),
         max_seq_len,
         KVBlockSize=kv_block_size,
-        Preshuffle=True,
+        Preshuffle=aiter_can_use_preshuffle_paged_mqa(),
```
[the remark this answers is not in the document]
Comment (13524182838) [reply python/sglang/srt/layers/attention/dsv4/indexer.py:168]: I resolved it in the latest version.

**Comment by amd-danli103** [comment]
@amd-bot ci-status

**Comment by 1am9trash** [comment]
Hi, @13524182838 
Thanks for the fix! 
Could you also share the exact test command / config you used to reproduce this accuracy issue? I checked yesterday's nightly CI and didn't see any accuracy regression alert.

CI status: [nightly-8-gpu-mi35x-deepseek-v4-pro-rocm720](https://github.com/sgl-project/sglang/actions/runs/29854963573/job/88716968735)

PS. 
Also with latest docker and cookbook setting, the acc seems correct.
docker: `rocm/sgl-dev:v0.5.15.post1-rocm720-mi35x-20260722 `
GSM8K:
```
100%|█████████████████████████| 1319/1319 [00:56<00:00, 23.37it/s]
Accuracy: 0.942
Invalid: 0.000
Latency: 56.441 s
Output throughput: 2066.782 token/s
```

File: python/sglang/kernels/jit/csrc/deepseek_v4/store.cuh [hunk 3]
```diff
@@ -114,7 +114,20 @@ __global__ void fused_store_indexer_cache(const __grid_constant__ FusedStoreCach
...
+    const int32_t dim_base = lane_id * kVecSize;
+    const int32_t token_tile_id = offset / kTile;
+    const int32_t token_in_tile = offset % kTile;
+    const int32_t col_tile_id = dim_base / kTile;
+    const int32_t col_in_tile = dim_base % kTile;
+    const int32_t value_offset =
+        token_tile_id * (kTile * 128) + col_tile_id * (kTile * kTile) + token_in_tile * kTile + col_in_tile;
+    static_cast<OutStorage*>(pointer::offset(page_ptr, value_offset))[0] = result;
+  } else {
+    static_cast<OutStorage*>(value_ptr)[lane_id] = result;
```
Comment (RolaoDenthu) [inline python/sglang/kernels/jit/csrc/deepseek_v4/store.cuh]: Same in this file. Not on hip path.

File: python/sglang/kernels/ops/attention/dsv4/attn.py [hunk 4]
```diff
@@ -31,7 +35,16 @@ def _jit_fused_store_module(
...
+    template_args = [input_dtype, index_dtype, page_size, is_arch_support_pdl()]
+    if name == "indexer":
+        # Only the indexer kernel has the preshuffle tile-size template param;
+        # FusedStoreCacheFlashMLAKernel does not take it.
+        template_args.append(
+            INDEXER_K_CACHE_PRESHUFFLE_TILE
+            if aiter_can_use_preshuffle_paged_mqa()
+            else 0
+        )
+    args = make_cpp_args(*template_args)
```
Comment (RolaoDenthu) [inline python/sglang/kernels/ops/attention/dsv4/attn.py]: Do we need changes here? fused_store_cache() in L67-88 routes hip to the triton_fused_store_cache(), so this function will never be used on hip.

**Comment by 13524182838** [comment]
> Hi, @13524182838 Thanks for the fix! Could you also share the exact test command / config you used to reproduce this accuracy issue? I checked yesterday's nightly CI and didn't see any accuracy regression alert.
> 
> CI status: [nightly-8-gpu-mi35x-deepseek-v4-pro-rocm720](https://github.com/sgl-project/sglang/actions/runs/29854963573/job/88716968735)
> 
> PS. Also with latest docker and cookbook setting, the acc seems correct. docker: `rocm/sgl-dev:v0.5.15.post1-rocm720-mi35x-20260722 ` GSM8K:
> 
> 
> 100%|█████████████████████████| 1319/1319 [00:56<00:00, 23.37it/s]
> Accuracy: 0.942
> Invalid: 0.000
> Latency: 56.441 s
> Output throughput: 2066.782 token/s
> 

I ran the longbench_v2. With the patch, you can get the score 0.316; Otherwise, it is only 0.079.
For gsm8k, the accuracy is similar with/without patch, because the context is too short to find the difference.
The server command is:

`
echo 0 > /proc/sys/kernel/numa_balancing
 
export SGLANG_OPT_FP4_MOE_K_ALIGN=128
export SGLANG_DSV4_INDEXER_K_CACHE_PRESHUFFLE=1
 
export SGLANG_DEFAULT_THINKING=1
export SGLANG_DSV4_REASONING_EFFORT=max
export SGLANG_OPT_DEEPGEMM_HC_PRENORM=false
export SGLANG_USE_AITER=1
export SGLANG_USE_ROCM700A=1
export SGLANG_OPT_USE_FUSED_COMPRESS=true
export SGLANG_HACK_FLASHMLA_BACKEND=unified_kv_triton
export SGLANG_OPT_FP8_WO_A_GEMM=false
export SGLANG_OPT_USE_JIT_INDEXER_METADATA=false
export SGLANG_OPT_USE_TOPK_V2=false
export SGLANG_OPT_USE_AITER_INDEXER=true
export SGLANG_OPT_USE_TILELANG_INDEXER=false
export SGLANG_OPT_USE_TILELANG_MHC_PRE=false
export SGLANG_OPT_USE_TILELANG_MHC_POST=false
export SGLANG_FP8_PAGED_MQA_LOGITS_TORCH=1
export SGLANG_OPT_USE_FUSED_COMPRESS_TRITON=true
export SGLANG_OPT_USE_MULTI_STREAM_OVERLAP=false
export SGLANG_ROCM_USE_MULTI_STREAM=false
export AITER_BF16_FP8_MOE_BOUND=0
export SGLANG_EAGER_INPUT_NO_COPY=false
export SGLANG_DP_USE_GATHERV=1
 
MODEL=/data/DeepSeek-V4-Pro
sglang serve \
   --model-path ${MODEL} \
   --trust-remote-code \
   --tp 8 \
   --disable-radix-cache \
   --attention-backend dsv4 \
   --max-running-request 256 \
   --page-size 256 \
   --mem-fraction-static 0.90 \
   --swa-full-tokens-ratio 0.1 \
   --chunked-prefill-size 8192 \
   --port 8000 \
   --disable-shared-experts-fusion \
   --tool-call-parser deepseekv4 \
   --reasoning-parser deepseek-v4 \
--speculative-algorithm EAGLE \
--speculative-num-steps 3 \
--speculative-num-draft-tokens 4 \
--speculative-eagle-topk 1 \
--dp 8 \
--enable-dp-attention \
--enable-prefill-delayer 
`

You can also run the ut test to reproduce the issue:

`
import torch
from aiter.ops.shuffle import shuffle_weight
from sglang.kernels.ops.attention.dsa.index_buf_accessor import (
    _set_k_and_s_triton,
    _use_aiter_preshuffle,
)
from sglang.srt.layers.attention.dsa.utils import aiter_can_use_preshuffle_paged_mqa
from sglang.kernels.ops.quantization.fp8_kernel import is_fp8_fnuz

HEAD = 128
FP8 = torch.float8_e4m3fnuz if is_fp8_fnuz() else torch.float8_e4m3fn


def check(page_size):
    dev = "cuda"
    torch.manual_seed(0)
    n = page_size  # fill one full page
    index_k = (torch.randn(n, HEAD, device=dev) * 4).to(FP8)
    scale = torch.ones(n, 1, device=dev, dtype=torch.float32)
    buf = torch.zeros(1, page_size * (HEAD + 4), device=dev, dtype=torch.uint8)
    loc = torch.arange(n, device=dev, dtype=torch.int64)  # page 0, slots 0..n-1

    # ---- REAL sglang store ----
    _set_k_and_s_triton(buf=buf, loc=loc, index_k=index_k, index_k_scale=scale, page_size=page_size)

    stored = buf[0, : page_size * HEAD].clone().cpu()             # fp8 region as written by sglang
    k_u8 = index_k.view(torch.uint8)                              # (n,128) row-major bytes
    exp_shuffle = shuffle_weight(k_u8.view(1, n, HEAD).clone(), layout=(16, 16)).reshape(-1).cpu()
    exp_legacy = k_u8.reshape(-1).cpu()

    m_sh = torch.equal(stored, exp_shuffle)
    m_lg = torch.equal(stored, exp_legacy)
    layout = "PRESHUFFLE(shuffle_weight)" if m_sh else ("LEGACY(row-major)" if m_lg else "UNKNOWN")
    print(f"page_size={page_size:3d} | REAL sglang store wrote: {layout:26s} "
          f"| == aiter gather layout(shuffle_weight): {m_sh}")
    return m_sh


def main():
    print(f"_use_aiter_preshuffle={_use_aiter_preshuffle} "
          f"aiter_can_use_preshuffle_paged_mqa()={aiter_can_use_preshuffle_paged_mqa()}\n")
    ok = True
    for ps in (16, 32, 64):
        ok &= check(ps)
    print()
    print("=> REAL store matches aiter preshuffle layout on all page sizes:", ok)
    print("   (expected True on ref/with-patch; False/LEGACY on orig_main/no-patch)")


if __name__ == "__main__":
    main()
`

File: python/sglang/kernels/ops/attention/dsv4/attn.py (hunk 4 above)
Comment (13524182838) [reply python/sglang/kernels/ops/attention/dsv4/attn.py]: > Do we need changes here? fused_store_cache() in L67-88 routes hip to the triton_fused_store_cache(), so this function will never be used on hip.

I removed this path in the second commit.

File: python/sglang/kernels/jit/csrc/deepseek_v4/store.cuh (hunk 3 above)
Comment (13524182838) [reply python/sglang/kernels/jit/csrc/deepseek_v4/store.cuh]: I removed the second path in the second commit.

**Review by RolaoDenthu — APPROVED** [review]
LGTM

**Comment by 1am9trash** [comment]
Hi, @13524182838 
May you fix the lint format error? This can be fixed by pre-commit check commit.
Other parts LGTM.

**Comment by 13524182838** [comment]
> Hi, @13524182838 May you fix the lint format error? This can be fixed by pre-commit check commit. Other parts LGTM.

Fixed the lint format error.

**Review by 1am9trash — APPROVED** [review]

File: python/sglang/srt/layers/attention/dsv4/indexer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (HaiShaw) [reply python/sglang/srt/layers/attention/dsv4/indexer.py:168]: solved in the aiter code, can you share the aiter pr / commit id?

**Review by HaiShaw — APPROVED** [review]

**Comment by HaiShaw** [comment]
CI clean
